src/server/object_services/object_server.js

```diff
@@ -1416,9 +1419,52 @@ async function find_object_md(req) {
         obj = await MDStore.instance().find_object_latest(req.bucket._id, req.rpc_params.key);
     }
     check_object_mode(req, obj, 'NO_SUCH_OBJECT');
+
+    // if part number was specified in the S3 Query
+    if (req.rpc_params.multipart_number) {
+        await calculate_multipart_range(req.rpc_params.multipart_number, obj);
+        if (!obj.num_multiparts) await calculate_num_multiparts(obj);
+    }
+
     return obj;
 }
 
+/**
+ * calculate (start, end) range of the specified multipart_number
+ *
+ * @param multipart_number integer in 1..10000 range
+ * @param {nb.ObjectMD} obj
+ */
+async function calculate_multipart_range(multipart_number, obj) {
+    // fetch the multiparts of this obj which are lte(<=) the requested part number
+    // TODO: store the range inside each multipart in _complete_object_multiparts() so we can read just the requested multipart here
+    const multiparts = await MDStore.instance().find_multiparts_of_object_lte_number(obj._id, multipart_number);
+    obj.multipart_start = 0;
+    obj.multipart_end = 0;
+    for (const mp of multiparts) {
+        if (mp.num < multipart_number) obj.multipart_start += mp.size;
+        obj.multipart_end += mp.size;
+    }
+}
+
+/**
+ * Lazily calculate the number of object's multiparts from the multiparts in the DB
+ * and populate the object num_multiparts field in object md DB
+ *
+ * @param {nb.ObjectMD} obj
+ */
+ async function calculate_num_multiparts(obj) {
+    // fetch the commited multiparts of this obj
+    const multiparts = await MDStore.instance().find_commited_multiparts_of_object(obj._id);
+    if (multiparts.length) {
+        obj.num_multiparts = multiparts.length;
+
+        // schedule it - forget it
+        // don't await the update, perform in "background"
+        MDStore.instance().update_object_by_id(obj._id, { num_multiparts: obj.num_multiparts });
```

Comment:
This looks like a point that is prone to race conditions or failures
I would have been worried about triggering in the background, is there a reason besides not slowing down the GET? 
If I'm not mistaken since this works on a particular object_id of finished uploads then there isn't risk of two parallel overrides setting different values and one of them winning the race condition and updating the invalid value of parts
I assume the above-mentioned case is not something that would happen since two overrides would have different object_ids, but I would verify if I'm missing something

Reply:
I'm not too worried about race conditions, but maybe to be on the safe side we can add the `await` here and not do it in the background. 
since this is only called when the part number is requested, it shouldn't affect the more common operations.

Reply:
@jeniawhite @dannyzaken  since 
- it is allowed for `num_multiparts` to be not initialized
- there is a single value  `num_multiparts` could take once initialized

I am not sure there is a race here. WDYT?

Reply:
I agree that there shouldn't be a race here, but I also understand Jenia's point - doing things in the background (especially updating DB) is bug-prone, and I think that in this case, it does not have a genuine performance improvement

Reply:
👍, going back before [that comment](https://github.com/noobaa/noobaa-core/pull/6526#issuecomment-852747220) then.